File: scrapers/scraper_goodbom.py
import requests  
from bs4 import BeautifulSoup  
import json  
import re

GOODBOM_URL = "https://www.goodbom.com.br/hortolandia/busca?q="
INPUT_FILE = "products.txt"
OUTPUT_JSON = "docs/prices_goodbom.json"  # arquivo separado
NUM_PRODUTOS = 3

def extrair_peso(nome):
    match = re.search(r"(\d+)\s*g", nome.lower())
    return int(match.group(1))/1000 if match else 1

def buscar_goodbom(produto):
    try:
        resp = requests.get(f"{GOODBOM_URL}{produto}", timeout=15)
        resp.raise_for_status()
        soup = BeautifulSoup(resp.text, "html.parser")
        encontrados = []

        for span in soup.find_all("span", class_="product-name")[:NUM_PRODUTOS]:
            a_tag = span.find_parent("a")
            preco_span = a_tag.find("span", class_="price") if a_tag else None
            if not preco_span: 
                continue

            nome = span.get_text(strip=True)
            if produto.lower() not in nome.lower():
                continue

            try:
                preco = float(preco_span.get_text(strip=True).replace("R$", "").split("/")[0].replace(",", ".").strip())
            except:
                continue

            peso_kg = extrair_peso(nome)
            encontrados.append({
                "supermercado": "Goodbom",
                "produto": nome,
                "preco": preco,
                "preco_por_kg": round(preco/peso_kg, 2)
            })

        return min(encontrados, key=lambda x: x["preco_por_kg"]) if encontrados else None
    except Exception as e:
        print(f"Erro Goodbom ({produto}): {e}")
        return None

def main():
    with open(INPUT_FILE, "r", encoding="utf-8") as f:
        produtos = [linha.strip() for linha in f if linha.strip()]

    resultados = []
    faltando = []

    for produto in produtos:
        item = buscar_goodbom(produto)
        if item:
            resultados.append(item)
        else:
            faltando.append(produto)

    with open(OUTPUT_JSON, "w", encoding="utf-8") as f:
        json.dump(resultados, f, ensure_ascii=False, indent=2)

    print("\nProdutos encontrados Goodbom:")
    for item in resultados:
        print(f"- {item['produto']}: R$ {item['preco']} | R$ {item['preco_por_kg']}/kg")

    total = sum(item["preco"] for item in resultados)
    print(f"\nTotal: R$ {total:.2f}")

    if faltando:
        print(f"\nProdutos não encontrados ({len(faltando)}): {', '.join(faltando)}")

if __name__ == "__main__":
    main()

File: scrapers/scraper_goodbom.js
const puppeteer = require("puppeteer");
const fs = require("fs");
const path = require("path");

const produtosTxtPath = path.join(__dirname, "..", "products.txt");
const outDir = path.join(__dirname, "..", "docs", "prices");

function normalizar(txt) {
  if (!txt) return "";
  return txt
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .trim();
}

function extrairPeso(nome) {
  const n = nome.toLowerCase();
  const matchG = n.match(/(\d+)\s*g/);
  if (matchG) return parseInt(matchG[1], 10) / 1000;
  const matchKg = n.match(/(\d+[.,]?\d*)\s*kg/);
  if (matchKg) return parseFloat(matchKg[1].replace(",", "."));
  const matchMl = n.match(/(\d+[.,]?\d*)\s*ml/);
  if (matchMl) return parseFloat(matchMl[1].replace(",", ".")) / 1000;
  const matchL = n.match(/(\d+[.,]?\d*)\s*l/);
  if (matchL) return parseFloat(matchL[1].replace(",", "."));
  return 1;
}

async function main() {
  const browser = await puppeteer.launch({ 
    headless: "new", 
    args: ["--no-sandbox", "--disable-setuid-sandbox"] 
  });
  const page = await browser.newPage();

  const produtos = fs.readFileSync(produtosTxtPath, "utf-8")
    .split("\n").map(l => l.trim()).filter(Boolean);

  const resultado = [];
  let totalEncontrados = 0;

  try {
    for (const [index, produto] of produtos.entries()) {
      const id = index + 1;
      let termoParaBusca = produto.replace(/\bkg\b/gi, "").replace(/\bg\b/gi, "").trim();
      const termoNorm = normalizar(termoParaBusca);

      console.log(`🔍 Buscando: ${termoParaBusca}`);

      try {
        await page.goto(
          `https://www.goodbom.com.br/hortolandia/busca?q=${encodeURIComponent(termoParaBusca)}`,
          { waitUntil: "networkidle2", timeout: 90000 }
        );

        // 1. ESPERA PELOS CARDS (Igual ao Tenda)
        await page.waitForSelector("span.product-name", { timeout: 15000 });
        
        // 2. SCROLL PARA ATIVAR PREÇOS
        await page.mouse.wheel({ deltaY: 400 });
        await new Promise(r => setTimeout(r, 1000));
      } catch (e) {
        console.log(`⚠️ Cards não apareceram para: ${termoParaBusca}`);
        continue;
      }

      const items = await page.evaluate(() => {
        const spans = Array.from(document.querySelectorAll("span.product-name"));
        return spans.slice(0, 15).map(span => {
          const nome = span.innerText.trim();
          const card = span.closest("a") || span.parentElement;
          
          // LÓGICA DE DUPLO SELETOR DE PREÇO
          let precoTxt = card.querySelector("span.price")?.innerText || 
                         card.querySelector(".sale-price")?.innerText || 
                         card.querySelector("[class*='price']")?.innerText || "0";
          
          return { nome, precoTxt };
        });
      });

      // 3. FILTRO POR RADICAL (3 letras)
      const filtrados = items.map(item => {
        const pTxt = item.precoTxt.replace("R$", "").replace(/\s/g, "").replace(",", ".").replace(/[^\d.]/g, "");
        return {
          nome: item.nome,
          preco: parseFloat(pTxt) || 0,
          peso_kg: extrairPeso(item.nome)
        };
      }).filter(item => {
        const nomeNorm = normalizar(item.nome);
        
        // Bloqueios de categoria
        if (!termoNorm.includes('suina') && nomeNorm.includes('suina')) return false;
        if (!termoNorm.includes('oleo') && nomeNorm.includes('oleo')) return false;

        // Todas as palavras da busca (radicais) devem estar no nome
        const palavrasBusca = termoNorm.split(" ").filter(w => w.length >= 3);
        const temTodas = palavrasBusca.every(pal => {
          const radical = pal.substring(0, 3);
          return nomeNorm.includes(radical);
        });

        return item.preco > 0 && temTodas;
      });

      if (filtrados.length > 0) {
        // Seleção por melhor preço por KG
        const ordenados = filtrados.map(item => ({
          ...item,
          preco_por_kg: parseFloat((item.preco / item.peso_kg).toFixed(2))
        })).sort((a, b) => a.preco_por_kg - b.preco_por_kg);

        const maisBarato = ordenados[0];
        totalEncontrados++;

        resultado.push({
          id,
          supermercado: "Goodbom",
          produto: maisBarato.nome,
          preco: maisBarato.preco,
          preco_por_kg: maisBarato.preco_por_kg
        });

        console.log(`✅ ${maisBarato.nome} - R$ ${maisBarato.preco.toFixed(2)}`);
      } else {
        console.log(`⚠️ Nenhum match válido para: ${produto}`);
      }
    }

    fs.writeFileSync(
      path.join(outDir, "prices_goodbom.json"),
      JSON.stringify(resultado, null, 2),
      "utf-8"
    );

    console.log(`📊 Finalizado GoodBom: ${totalEncontrados}/${produtos.length}`);
  } catch (err) {
    console.error("❌ Erro fatal GoodBom:", err.message);
  } finally {
    await browser.close();
  }
}

main();
          
